src/DataStructures/Tensor/Expressions/DataTypeSupport.hpp

```diff
@@ -0,0 +1,387 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+/// \file
+/// Defines which types are allowed, whether operations which certain types are
+/// allowed, and other type-specific properties and configuration for
+/// `TensorExpression`s
+///
+/// \details
+/// To add support for a data type, modify the templates in this file and the
+/// arithmetic operator overloads as necessary. Then, add tests as appropriate.
+
+#pragma once
+
+#include <cstddef>
+#include <limits>
+#include <type_traits>
+
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/Expressions/TensorExpression.hpp"
+#include "DataStructures/VectorImpl.hpp"
+
+namespace tenex {
+template <typename DataType>
+struct NumberAsExpression;
+
+namespace detail {
+/// \brief Whether or not `TensorExpression`s supports using a given type as a
+/// numeric term
+///
+/// \details
+/// To make it possible to use a new numeric data type as a term in
+/// `TensorExpression`s, add the type to this alias and adjust other templates
+/// in this file, as necessary.
+///
+/// \tparam X the arithmetic data type
+template <typename X>
+using is_supported_number_datatype =
```

Comment:
Since I have a boolean expression `std::is_same_v<...> or std::is_same_v<...>`, I think I think I can't get rid of `std::bool_constant` because I need to wrap the expression with it. However, I imagine I can inherit from `std::bool_constant`. Does the following work for you or did you have something else in mind that I'm not following?:
```
template <typename X>
struct is_supported_number_datatype
    : std::bool_constant<std::is_same_v<X, double> or
                         std::is_same_v<X, std::complex<double>>> {};
```
Alternatively, I could define a specialization for each and do this instead where we _can_ inherit off of `std:is_same`:
```
template <typename X>
struct is_supported_number_datatype;

template <typename X>
struct is_supported_number_datatype : std::false_type {};
template <>
struct is_supported_number_datatype : std::is_same<X, double>> {};
template <>
struct is_supported_number_datatype : std::is_same<X, std::is_complex<double>>> {};
```

Let me know if you have a preference or a different idea

Reply:
https://en.cppreference.com/w/cpp/types/disjunction

Reply:
Wow, never seen this but this is a really nice thing to know about